### ISSUE TITLE
Struggles with Extending Schemathesis 

### ISSUE THREAD

**Issue by barrett-schonefeld** [kiwicom/schemathesis#442]
### Intro

Hey, folks, I am experimenting with extending Schemathesis with additional validation rules. I created the following module. (Note: `body_no_content_type` function is only relevant for an invalid request with a request body but no Content-Type header. I will focus on finding a way to send this invalid request once I understand how to use the `runner`.)

    from src import schemathesis
    from src.schemathesis import runner, models
    
    def body_no_content_type(response, case: models.Case):
        if case.body and 'content-type' not in case.headers:
            assert response.status_code == '415'
    
    runner.execute(
        '<path to my OpenAPI description>',
        checks=[body_no_content_type],
        loader=schemathesis.from_path
    )

### Problem

I'm struggling because I don't know what I need to do to run Schemathesis (with all its builtin request generation, validation, and error reporting) with this `body_no_content_type` check added to the set of validation functions.

### Attempted Solutions

(Minimal Python experience, so bear with me if the solution is obvious and Python/Pytest specific.)

My initial attempt was to simply run the script as a Python program. However, this approach produces no test result output and makes no requests to the server. I supply the URL of the server in the `servers` field of my OpenAPI description. I found in the code that Schemathesis grabs the first URL in this list as the URL for the API. However, I'm wondering if the runner somehow bypasses this step and needs the URL as an argument.

Next I tried running the script manually through the Python shell to see if any specific section of the script was failing. After running the `runner.execute()` command in the Python shell, I got the output `<src.schemathesis.models.TestResultSet object at 0x1069b0450>`. Still, no requests were sent to the server.

Note: after creating this issue, I added a path to my API that returns the API definition, and using the following command:

    runner.execute(
        'http://localhost:3000/apidef.json',
        checks=[body_no_content_type],
        loader=schemathesis.from_uri
    )

Schemathesis did send a request for each path in the API description. However, this did not produce a report of the results.

### Purpose

My team is considering Schemathesis for an API endpoint validation project. We need to extend Schemathesis with our API handbook specific rules, and we are implementing one of those rules to see if Schemathesis meets our extensibility needs.

**Comment by Stranger6667** [comment]
Hi @barrett-schonefeld 
Let's take a look! :) It seems to me that our `execute` is not documented enough to be easily used and does not give clues when some configuration is missing. It is definitely something we want to improve :)

> However, I'm wondering if the runner somehow bypasses this step and needs the URL as an argument.

You are right, as far as I see you need to add `loader_options={"base_url": "<base URL of your API>"}` to the `execute` call. Let me know if it will work for you.

As far as I remember automatic detection was implemented some time ago, but then I removed it because often such local files have URLs to production services and it might generate some unwanted load on these services. It is checked in the CLI, but not checked in `runner`.

Also, I forgot to remove `execute` :( it wasn't intended to be there anymore.

Initially `runner` was blocking, and there was no way to stop it, after it it was changed to a generator to display its progress dynamically (it may be not only CLI, but any place, e.g. it can be served via some websocket and displayed in a web interface). 

So, the function which is intended to be used is `prepare`, but I didn't get back to documenting it properly.

It is a generator that should be used like this:

```
runner = prepare(...)
for event in runner:
    do_something(event)
```

However, if you need CLI output that Schemathesis provides together with your own checks it could be done like this:

1. Create a `.py` file with your check

```
import schemathesis

@schemathesis.register_check
def body_no_content_type(response, case: models.Case):
    if case.body and 'content-type' not in case.headers:
        assert response.status_code == '415'
```

2. Include it in a CLI command:

```schemathesis --pre-run module.with.checks run -c all --base-url http://example.com /local/path/to/schema.json```

This command will execute your module with `body_no_content_type` check (`--pre-run` part) and will run all checks (`-c all` is a shortcut to include all available checks) against the API at the provided `base-url`

I already have a WIP branch with cleaning up the runner interface, and will address these issues there.

Let me know if you have any additional questions and if my suggestions work in your case as intended

**Comment by Stranger6667** [comment]
In other words, runner only generates events, that are handled later by CLI.
Btw, there is a bit of documentation that might be helpful https://github.com/kiwicom/schemathesis#pre-run-cli-hook

**Comment by barrett-schonefeld** [comment]
Thanks for the help!

I will give your suggestion a try and let you know how it goes.

**Comment by barrett-schonefeld** [comment]
I'm getting an error on the `__import__` when Schemathesis tries to import the hook module.
    
    ../site-packages/schemathesis/cli/__init__.py", line 245, in load_hook
        __import__(module_name)
    ModuleNotFoundError: No module named 'hook.py'

Where should these hook files live? I have the hooks module in my current working directory, and while trying to get the import statement to work, I added a copy of the hook in the `../schemathesis/cli` directory where the actual import statement lives. I imagine the hook needs to live somewhere where Schemathesis is in scope because the first thing we do is `import schemathesis`.

Also, in the example you provided `schemathesis --pre-run module.with.checks run -c all --base-url http://example.com /local/path/to/schema.json`, why do you use `module.with.checks` as the example name? Does the `.`-separated name hint at how we should format this argument?

**Comment by Stranger6667** [comment]
Hook module could be in any place, discoverable in the current `PYTHONPATH`. Schemathesis accepts an importable string as a path. For example, if you have "hook.py" in your current directory, then it should be "hook" (in the same way you could import this module with `import hook`). If I create `hook.py` in Schemathesis repo root, then I can use it in this way:

```
PYTHONPATH=`pwd` schemathesis --pre-run hook run http://example.com/schema.json
```

Let me know if this works for you :)

**Comment by barrett-schonefeld** [comment]
Great, solved the import error!

The tests are running now, but the hook is not being executed with the checks. I tested this by changing the hook behavior to `assert False` and yet some of the tests still pass with the command:

    PYTHONPATH=`pwd` schemathesis --pre-run src.schemathesis.handbook_rules.hook run <path to API description>  -c all --base-url http://localhost:3000

Then I tried following the example from the hooks section of the README by using `-c body_no_content_type` try only running the hook.

    PYTHONPATH=`pwd` schemathesis --pre-run src.schemathesis.handbook_rules.hook run <path to API description> -c body_no_content_type --base-url http://localhost:3000

When I did this, I received the following error:

    Error: Invalid value for "--checks" / "-c": invalid choice: body_no_content_type. (choose from
    not_a_server_error, status_code_conformance, content_type_conformance,
    response_schema_conformance, all)

**Comment by Stranger6667** [comment]
Do you have your hook function wrapped with `@schemathesis.register_hook` ?

Here is my `hook.py` in the schemathesis root:

```
import schemathesis

@schemathesis.register_check
def body_no_content_type(response, case):
    if case.body and 'content-type' not in case.headers:
        assert response.status_code == '415'
```

And I run it with wrong `-c` value to check the error message:

```
⇒  PYTHONPATH=`pwd` schemathesis --pre-run hook run -c wrong --validate-schema=false http://127.0.0.1:8081/swagger.yaml 
Usage: schemathesis run [OPTIONS] SCHEMA
Try 'schemathesis run -h' for help.

Error: Invalid value for '--checks' / '-c': invalid choice: wrong. (choose from not_a_server_error, status_code_conformance, content_type_conformance, response_schema_conformance, all, **body_no_content_type**)
```

And it is there, in the last position, so it is registered.

**Comment by barrett-schonefeld** [comment]
Got it!

I imported the wrong Schemathesis in my hook module. I was importing the Schemathesis source code.

The hook is running beautifully now.

**Comment by Stranger6667** [comment]
Great :) Let me know if I can do anything to make Schemathesis more flexible for your case.

I'll leave this issue open until the mentioned cleanups will be released

**Comment by barrett-schonefeld** [comment]
Sounds great, thank you for the help!

**Comment by Stranger6667** [comment]
The change I mentioned is implemented in 0a1412968fcdc2be3d589bf08daa49bb62038812
I also updated the docs to make them more clear on extending Schemathesis
